tracing-subscriber/src/fmt/time.rs

```diff
@@ -268,3 +268,210 @@ where
+            Ok(duration) => {
+                debug_assert!(duration.as_secs() <= i64::MAX as u64);
+                (duration.as_secs() as i64, duration.subsec_nanos())
+            }
+            Err(error) => {
+                let duration = error.duration();
+                debug_assert!(duration.as_secs() <= i64::MAX as u64);
+                let (secs, nanos) = (duration.as_secs() as i64, duration.subsec_nanos());
+                if nanos == 0 {
+                    (-secs, 0)
+                } else {
+                    (-secs - 1, 1_000_000_000 - nanos)
+                }
+            }
+        };
+
+        // 2000-03-01 (mod 400 year, immediately after feb29
+        const LEAPOCH: i64 = 946_684_800 + 86400 * (31 + 29);
+        const DAYS_PER_400Y: i32 = 365 * 400 + 97;
+        const DAYS_PER_100Y: i32 = 365 * 100 + 24;
+        const DAYS_PER_4Y: i32 = 365 * 4 + 1;
+        static DAYS_IN_MONTH: [i8; 12] = [31, 30, 31, 30, 31, 31, 30, 31, 30, 31, 31, 29];
+
+        // Note(dcb): this bit is rearranged slightly to avoid integer overflow.
+        let mut days: i64 = (t / 86_400) - (LEAPOCH / 86_400);
+        let mut remsecs: i32 = (t % 86_400) as i32;
+        if remsecs < 0i32 {
+            remsecs += 86_400;
+            days -= 1
+        }
+
+        let mut qc_cycles: i32 = (days / i64::from(DAYS_PER_400Y)) as i32;
+        let mut remdays: i32 = (days % i64::from(DAYS_PER_400Y)) as i32;
+        if remdays < 0 {
+            remdays += DAYS_PER_400Y;
+            qc_cycles -= 1;
+        }
+
+        let mut c_cycles: i32 = remdays / DAYS_PER_100Y;
+        if c_cycles == 4 {
+            c_cycles -= 1;
+        }
+        remdays -= c_cycles * DAYS_PER_100Y;
+
+        let mut q_cycles: i32 = remdays / DAYS_PER_4Y;
+        if q_cycles == 25 {
+            q_cycles -= 1;
+        }
+        remdays -= q_cycles * DAYS_PER_4Y;
+
+        let mut remyears: i32 = remdays / 365;
+        if remyears == 4 {
+            remyears -= 1;
+        }
+        remdays -= remyears * 365;
+
+        let mut years: i64 = i64::from(remyears)
+            + 4 * i64::from(q_cycles)
+            + 100 * i64::from(c_cycles)
+            + 400 * i64::from(qc_cycles);
+
+        let mut months: i32 = 0;
+        while i32::from(DAYS_IN_MONTH[months as usize]) <= remdays {
+            remdays -= i32::from(DAYS_IN_MONTH[months as usize]);
+            months += 1
+        }
+
+        if months >= 10 {
+            months -= 12;
+            years += 1;
+        }
+
+        DateTime {
+            year: years + 2000,
+            month: (months + 3) as u8,
+            day: (remdays + 1) as u8,
+            hour: (remsecs / 3600) as u8,
+            minute: (remsecs / 60 % 60) as u8,
+            second: (remsecs % 60) as u8,
+            nanos,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::i32;
+    use std::time::{Duration, UNIX_EPOCH};
+
+    use super::*;
+
+    #[test]
+    fn test_datetime() {
```

Comment:
It looks like this test is failing on Windows: https://github.com/tokio-rs/tracing/pull/807/checks?check_run_id=861424835#step:4:891

Reply:
I tried testing on a windows machine, and found that it was these test cases that were causing to panic.
https://github.com/Pothulapati/tracing/blob/1c7c70b91ba7504aa486f69209d609544eb17a4e/tracing-subscriber/src/fmt/time.rs#L463 (and this is not the conversion logic to DateTime but it was in the `case` helper closure) as they were using `i64::Max` or`i64::Min` with `UNIX_EPOCH + Duration::new(secs as u64, micros * 1_000)` and `impl Add<Duration> for SystemTime` panics if the resulting point in time cannot be represented by the underlying data structure as per https://doc.rust-lang.org/std/time/struct.SystemTime.html#panics

Wonder why the underlying data structure is different in windows? :thinking: Also, As the duration in the test case is pretty high i.e giving the time `+292277026596-12-04T15:30:07.000000Z`. Should it matter? 

Thanks

Reply:
Hmm. I think timestamps in the year 292277026596 are _pretty_ unlikely to occur any time soon. However, there are situations where weird timestamps can occur other than actually being in that year --- the system clock might be misconfigured or something.

I think it would be better to handle this case a little more gracefully than wrapping. We could probably add a check before doing that add, and if it would wrap, we could just print "Invalid timestamp" or something, rather than panicking. 

For the short term, I think it would be okay to disable the specific tests which use the really high year value on Windows. We can do that by adding a `cfg` attribute:
```rust
#[cfg(not(target_os = "windows"))] 
{
        case("+292277026596-12-04T15:30:07.000000Z", i64::MAX, 0);
        case("+292277026596-12-04T15:30:06.000000Z", i64::MAX - 1, 0);
        case("-292277022657-01-27T08:29:53.000000Z", i64::MIN + 1, 0);
}
```
If we do that, it would be good to have a comment explaining why those tests are disabled on Windows.

Reply:
Probably also worth ccing @danburkert — he might want to be aware that the upstream version of this test doesn't work with Windows timestamps, too. :)